src/Build/BackEnd/Node/MSBuildClient.cs

```diff
@@ -0,0 +1,661 @@
+            SendBuildCommand(commandLine, _nodeStream);
+
+            InitializeAsyncPacketThread();
+
+            var waitHandles = new WaitHandle[] {_packetPumpShutdownEvent, _packetReceivedEvent };
+
+            // Get the current directory before doing any work. We need this so we can restore the directory when the node shutsdown.
+            while (!_buildFinished)
+            {
+                int index = WaitHandle.WaitAny(waitHandles);
+                switch (index)
+                {
+                    case 0:
+                        CommunicationsUtilities.Trace($"MSBuild client error: packet pump unexpectedly shutted down.");
+                        _exitResult.MSBuildClientExitType = MSBuildClientExitType.Unexpected;
+                        return _exitResult;
+
+                    case 1:
+                        while (_receivedPacketsQueue.TryDequeue(out INodePacket? packet) && (!_buildFinished))
+                        {
+                            if (packet != null)
+                            {
+                                try
+                                {
+                                    HandlePacket(packet);
+                                }
+                                catch (Exception ex)
+                                {
+                                    CommunicationsUtilities.Trace($"MSBuild client error: problem during packet handling occured. {ex.Message}");
+                                    _exitResult.MSBuildClientExitType = MSBuildClientExitType.Unexpected;
+                                    return _exitResult;
+                                }
+                            }
+                        }
+
+                        break;
+                }
+            }
+
+            CommunicationsUtilities.Trace("Build finished.");
+            return _exitResult;
+        }
+
+        /// <summary>
+        /// Launches MSBuild server. 
+        /// </summary>
+        /// <returns> Whether MSBuild server was started successfully.</returns>
+        private bool LaunchMSBuildServer()
+        {
+            string serverLaunchMutexName = $@"Global\server-launch-{_pipeName}";
+            using var serverLaunchMutex = ServerNamedMutex.OpenOrCreateMutex(serverLaunchMutexName, out bool mutexCreatedNew);
+            if (!mutexCreatedNew)
+            {
+                // Some other client process launching a server and setting a build request for it. Fallback to usual msbuild app build.
+                CommunicationsUtilities.Trace("Another process launching the msbuild server, falling back to former behavior.");
+                _exitResult.MSBuildClientExitType = MSBuildClientExitType.ServerBusy;
+                return false;
+            }
+
+            string[] msBuildServerOptions = new[] {
+                _dllLocation,
+                "/nologo",
+                "/nodemode:8"
+            }.ToArray();
```

Comment:
Is it necessary to invoke `ToArray` here? The result of `new[]` should already represent an array of strings.